doc/HowToUseMeson.md

```diff
@@ -0,0 +1,234 @@
+# How to use Meson as build system in your Python-Cython app
+
+----------
+
+<p align="center">
+    <a href="https://mesonbuild.com/index.html">
+    <img width="50" 
+         height="50" 
+         src="https://mesonbuild.com/assets/images/meson_logo.png" alt="meson source">
+    </a>
+</p>
+
+# Table of contents
+
+1. [Introduction](#introduction)
+2. [Project structure](#paragraph1)
+3. [Top level meson file](#paragraph2)
+    1. [Project settings](#subparagraph21)
+    2. [Useful system checks](#subparagraph22)
+4. [Meson + Python](#paragraph3)
+5. [Meson + Cython](#paragraph4)
+5. [How to compile apps, which are using mc_lib with meson](#paragraph5)
+
+---------
+## <a name="introduction"></a> Introduction 
+
+Some introduction. I'll add it later
+
+
+## <a name="paragraph1"></a> Project structure 
+
+All `meson.build` files connected between each other, it means,
+that variables deaclared in upper `meson.build` are visible
+from every subdirectorys `meson.build` files.
+
+The order of subdir() links is important for project. 
+
+Example:
+```
+project
+│   file001.pyx
+│   meson.build (A) 
+│   
+└───subdir1
+│   │   file011.pyx
+│   │   meson.build (B)
+│   │
+│   └───subdir1_1
+│       │   file111.pyx
+│       │   file112.pxd
+│       │   meson.build (C)
+│   
+└───subdir2
+    │   file021.pyx
+    │   meson.build (D)
+```
+
+`meson.build (A)`:
+
+```meson
+subdir('subdir1')
+subdir('subdir2')
+```
+Variables declared in `subdir1` are visible in `subdir2`, but not vice versa.
+
+
+## <a name="paragraph2"></a> Top level `meson.build`
+
+
+### <a name="subparagraph21"></a> Project settings 
+There should be declared:
+1. Project name,
+2. Programming languages(meson trying to find a suitable compiler for named languages ),
+3. Suitable meson version,
+4. App version,
+5. Licenses,
+6. Etc.
+
+Cython files compiles in order: `.pyx` &rarr; `.c` or `.cpp` &rarr; `.so`.
+So there must be declared at least 2 languages: `cpp` or `c` and `cython`.
+
+
+### <a name="subparagraph22"></a> Useful system checks 
+For apps which are using `cython` can be added manual check
+using `find_program` function:
+```meson
+cython = find_program('cython', required : true)
+if not cython.found()
+  error('MESON_BUILD_FAILED: Cython3 not found.')
+endif
+```
+
+Also, the `python` installation must be checked too. 
+Default `pyhton` dependencies saved into variable
+to insert them into every compile target
+```meson
+py_mod = import('python')
+py3 = py_mod.find_installation()
+py3_dep = py3.dependency()
+```
+As for `MC_lib`, the most part of it uses `numpy` API,
+so numpy declared as dependency, by its location:
+```meson
+incdir_numpy = run_command(py3,
+  ['-c', 'import os; os.chdir(".."); import numpy; print(numpy.get_include())'],
+  check : true
+).stdout().strip()
+
+inc_np = include_directories(incdir_numpy)
+```
```

Comment:
Side note: this is a bit cumbersome. I plan to upstream this into Meson, so you can just do `dependency('numpy')`.

Reply:
That'd be great! Just numpy or as a general pattern? The latter might be useful downstream, too.

Reply:
Just for NumPy. `packagename.get_include()` is not common enough that it should be standardized I think. The only packages I can think of that do this are numpy, pybind11 and this repo.

Reply:
Also, numpy has multiple paths that may be needed, like for `libnpyrand` or `fortranobject.c`

Reply:
(this repo took it straight from numpy) I just don't know of a better pattern. Or at least as long as the build system was in python, so interpreting python code in setup.py was natural, that is :-)

Reply:
Yes, I'm not sure if there is anything better. There are just not many Python projects that offer a C or Cython API. If you do it the standard non-Python way, then headers get installed completely separate from executable code. So for a virtualenv, it'd be `envname/include` instead of `envname/lib/python3.X/site-packages/`. I've just never seen that done and I suspect it'd be quite hard to get right.

Reply:
Yup, this settles it --- minor copy-pastable magic is way better than separate `include/`, `lib/` and `bin/` directories IMO! 

(As long as the hard work is done once and we all can copy-paste it)